Firestore/core/test/firebase/firestore/local/serializer_test.cc

```diff
@@ -0,0 +1,148 @@
+using firebase::firestore::testutil::Doc;
+using firebase::firestore::util::Status;
+using firebase::firestore::util::StatusOr;
+using google::protobuf::util::MessageDifferencer;
+
+// TODO(rsgowman): This is copied from remote/serializer_tests.cc. Refactor.
+#define EXPECT_OK(status) EXPECT_TRUE(StatusOk(status))
+
+// Fixutre for running the local serializer tests.
+class SerializerTest : public ::testing::Test {
+ public:
+  SerializerTest()
+      : remote_serializer(kDatabaseId), serializer(remote_serializer) {
+    msg_diff.ReportDifferencesToString(&message_differences);
+  }
+
+  const DatabaseId kDatabaseId{"p", "d"};
+  remote::Serializer remote_serializer;
+  local::Serializer serializer;
+
+  void ExpectRoundTrip(const MaybeDocument& model,
+                       const firestore::client::MaybeDocument& proto,
+                       MaybeDocument::Type type) {
+    // First, serialize model with our (nanopb based) serializer, then
+    // deserialize the resulting bytes with libprotobuf and ensure the result is
+    // the same as the expected proto.
+    ExpectSerializationRoundTrip(model, proto, type);
+
+    // Next, serialize proto with libprotobuf, then deserialize the resulting
+    // bytes with our (nanopb based) deserializer and ensure the result is the
+    // same as the expected model.
+    ExpectDeserializationRoundTrip(model, proto, type);
+  }
+
+  /**
+   * Checks the status. Don't use directly; use one of the relevant macros
+   * instead. eg:
+   *
+   *   Status good_status = ...;
+   *   ASSERT_OK(good_status);
+   *
+   *   Status bad_status = ...;
+   *   EXPECT_NOT_OK(bad_status);
+   */
+  // TODO(rsgowman): This is copied from remote/serializer_tests.cc. Refactor.
+  testing::AssertionResult StatusOk(const Status& status) {
+    if (!status.ok()) {
+      return testing::AssertionFailure()
+             << "Status should have been ok, but instead contained "
+             << status.ToString();
+    }
+    return testing::AssertionSuccess();
+  }
+
+  // TODO(rsgowman): This is copied from remote/serializer_tests.cc. Refactor.
+  template <typename T>
+  testing::AssertionResult StatusOk(const StatusOr<T>& status) {
+    return StatusOk(status.status());
+  }
+
+ private:
+  void ExpectSerializationRoundTrip(
+      const MaybeDocument& model,
+      const firestore::client::MaybeDocument& proto,
+      MaybeDocument::Type type) {
+    EXPECT_EQ(type, model.type());
+    std::vector<uint8_t> bytes = EncodeMaybeDocument(&serializer, model);
+    firestore::client::MaybeDocument actual_proto;
+    bool ok = actual_proto.ParseFromArray(bytes.data(), bytes.size());
+    EXPECT_TRUE(ok);
+    EXPECT_TRUE(msg_diff.Compare(proto, actual_proto)) << message_differences;
+  }
+
+  void ExpectDeserializationRoundTrip(
+      const MaybeDocument& model,
+      const firestore::client::MaybeDocument& proto,
+      MaybeDocument::Type type) {
+    size_t size = proto.ByteSizeLong();
+    std::vector<uint8_t> bytes(size);
```

Comment:
Optional: similarly to another PR, I think you can do away with storing `size` as a local variable:
```cpp
std::vector<uint8_t> bytes(proto.ByteSizeLong());
bool status = proto.SerializeToArray(bytes.data(), bytes.size());
```

Reply:
Done.